src/devices.jl

```diff
@@ -52,19 +67,11 @@ function device()
     if CUDA_jl.functional()
         return CUDADevice()
     else
-        return CPUDevice()
+        nthreads = get(ENV, "JULIA_NUM_THREADS", 1)
+        return nthreads == 1 ? CPUSingleThreaded() : CPUMultiThreaded()
     end
 end
 
-"""
-    ClimaComms.device(object)
-
-Return the appropriate device for the specified object.
-"""
-device(::Array) = CPUDevice()
```

Comment:
Needing to remove the device makes me wonder if we should instead specialize on the threaded type. For example,

```julia
abstract type AbstractThreading end
struct SingleThreaded <: AbstractThreading end
struct MultiThreaded <: AbstractThreading end

abstract type AbstractDevice end
struct CPUDevice{T<:AbstractThreading} <: AbstractDevice end
struct CUDADevice <: AbstractDevice end
```

Specialization can be more brittle however. What do you think @simonbyrne?

Reply:
I don't know. Lets see how this pans out and we can re-evaluate.